meegkit/utils/trca.py

```diff
@@ -0,0 +1,157 @@
+                        h_int], [sigma_ci_lower, sigma_ci_upper]
+
+
+def itr(n, p, t):
+    """Compute information transfer rate (ITR).
+
+    Inputs
+    ------
+    n : int
+        Number of targets.
+    p : float
+        Target identification accuracy (0 <= p <= 1).
+    t : float
+        Average time for a selection (s).
+
+    Returns
+    -------
+    itr : float
+        Information transfer rate [bits/min]
+
+    References
+    ----------
+    .. [1] M. Cheng, X. Gao, S. Gao, and D. Xu,
+        "Design and Implementation of a Brain-Computer Interface With High
+        Transfer Rates", IEEE Trans. Biomed. Eng. 49, 1181-1186, 2002.
+
+    """
+    itr = 0
+
+    if (p < 0 or 1 < p):
+        raise ValueError('Accuracy need to be between 0 and 1.')
+    elif (p < 1 / n):
+        raise ValueError('ITR might be incorrect because accuracy < chance')
+        itr = 0
+    elif (p == 1):
+        itr = np.log2(n) * 60 / t
+    else:
+        itr = (np.log2(n) + p * np.log2(p) + (1 - p) *
+               np.log2((1 - p) / (n - 1))) * 60 / t
+
+    return itr
+
+
+def filterbank(eeg, fs, idx_fb):
+    """Filter bank design for decomposing EEG data into sub-band components.
+
+    Parameters
+    ----------
+    eeg : np.array, shape=(trials, channels, samples)
+        Training data.
+    fs : int
+        Sampling frequency of the data.
+    idx_fb : int
+        Index of filters in filter bank analysis
+
+    Returns
+    -------
+    y: np.array, shape=(n_trials, n_chans, n_samples)
+        Sub-band components decomposed by a filter bank.
+
+    """
+    if(eeg.ndim == 3):
+        num_chans = eeg.shape[1]
+        num_trials = eeg.shape[0]
+
+    elif(eeg.ndim == 2):  # Testdata come with only one trial at the time
+        num_chans = eeg.shape[0]
+        num_trials = 1
+
+    fs = fs / 2
+
+    passband = [6, 14, 22, 30, 38, 46, 54, 62, 70, 78]
+    stopband = [4, 10, 16, 24, 32, 40, 48, 56, 64, 72]
+    Wp = [passband[idx_fb] / fs, 90 / fs]
+    Ws = [stopband[idx_fb] / fs, 100 / fs]
+    # Chebyshev type I filter order selection.
+    N, Wn = cheb1ord(Wp, Ws, 3, 40)
+
+    # Chebyshev type I filter design
+    B, A = cheby1(N, 0.5, Wn, btype="bandpass")
+
+    y = np.zeros(eeg.shape)
+    if num_trials == 1:  # For testdata
+        for ch_i in range(num_chans):
+            try:
+                y[ch_i, :] = filtfilt(
+                    B, A, eeg[ch_i, :], axis=0, padtype='odd',
+                    padlen=3 * (max(len(B), len(A)) - 1))
+            except BaseException:
+                print(num_chans)
+    else:
+        for trial_i in range(num_trials):  # Filter each trial sequentially
+            for ch_i in range(num_chans):  # Filter each channel sequentially
+                # the arguments 'axis=0, padtype='odd',
+                # padlen=3*(max(len(B),len(A))-1)' correspond to Matlab
+                # filtfilt : https://dsp.stackexchange.com/a/47945
+                y[trial_i, ch_i, :] = filtfilt(
```

Comment:
@gferraro2019 I'm a bit surprised by the use of a non-causal filter (`filtfilt`) here in the context of a BCI, where the data is supposed to be acquired in real-time. Usually for BCIs we would use causal filtering. Do you have any comment on that?

Reply:
I'm sorry I don't have an answer for that.